### PR TITLE
Fix indexing, reindex on all-sparse SparseArray.

### LINKED ISSUE #35286: REGR: reindex with sparse data
See discussion at https://github.com/pandas-dev/pandas/pull/34158#issuecomment-636641583 and below. There was some discussion on the PR, but I think we didn't yet create an issue to track this regression.

The PR introduced a regression in `reindex` when having sparse data. Based on the linked discussion, there doesn't seem to direct / easy solution (but I didn't look into it again). So we might want to revert the PR for 1.1 until we figure out a correct solution.

cc @TomAugspurger

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/35286.

Also added a regression tests for the issue reported there.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Looks like https://github.com/pandas-dev/pandas/pull/34908 built on the buggy functionality. I may need to revert it as well.

File: pandas/tests/extension/test_sparse.py [hunk 1]
```diff
@@ -41,11 +41,6 @@ def data_for_twos(request):
     return SparseArray(np.ones(100) * 2)
 
 
-@pytest.fixture(params=[0, np.nan])
-def data_zeros(request):
-    return SparseArray(np.zeros(100, dtype=int), fill_value=request.param)
-
```
Comment (TomAugspurger) [inline pandas/tests/extension/test_sparse.py:47]: I don't think this was doing anything on master.

File: pandas/tests/extension/base/getitem.py [hunk 2]
```diff
@@ -399,31 +399,3 @@ def test_item(self, data):
 
         with pytest.raises(ValueError, match=msg):
             s.item()
-
-    def test_boolean_mask_frame_fill_value(self, data):
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/getitem.py:403]: I noticed that these tests were passing regardless of the change, so I'm not sure they were testing what we expected. I've left them removed.

**Review by TomAugspurger — COMMENTED** [review]
a089991e85a183202ebb08f2cedddac021f541b2 hopefully has the proper fix.

**Review by jreback — APPROVED** [review]
lgtm. wasn't clear if you had additional changes. merge when ready.

**Comment by TomAugspurger** [comment]
Thanks, I think we're good here.

**Pull request merged**